### ISSUE TITLE
Bug: Beat Layer no longer displayed + unable to save and close window after opening file

### ISSUE THREAD

**Issue by m-k94** [TimeLineAnnotator/desktop#205]
I was working on a file with different kinds of annotations and a beat/measure layer for a few days and everything was normal. But now when I open it, I just see the harmony and form annotations, but not the beat and marker layers that I had before. When I open the TLA file in an editor, all the information appears to still be there. And it even does get properly displayed in the Web-App: https://tilia-ad98d.web.app/viewer/66 

It also has some other strange side effects on the desktop version: Once the file is opened, I can interact with it, but I am unable to save it, open a different file or close the window normally. I have to do use task manager. Other files I created - so far - do not cause such issues.

I _think_ the bug might have started when I pressed Ctrl+z to undo an action. That unexpectedly caused a lot of stuff to disappear all at once, and when I reopened everything to start over from the last save, it started behaving as described above.

**Comment by FelipeDefensor** [comment]
Could you send me the .tla?

I'm guessing that something went wrong when undoing leading to an invalid timeline state. The overhaul that Loading a file in the web app happens slightly differently, so the faulty state may not have consequences there.

I would imagine that downloading the file you uploaded into the web app should give you a healthy file.

**Comment by FelipeDefensor** [comment]
The undoing/redoing process already went through an overhaul in the `optimize-restore-state` branch, so the changes there might have prevented whatever lead to the faulty state in the first place.

In any case, the app should be resilient to things going wrong when undoing/redoing. Perhaps it could check if the timeline state after the restore is equal to the desired state (the previous one, when undoing), and, if not, revert to the last known healthy state. Will look into that for v0.5.

**Comment by m-k94** [comment]
Thanks for the quick response! I tried downloading the file, but it causes the app to fully crash:

![image](https://github.com/user-attachments/assets/653b0c08-3730-4793-af5b-3f74f402a6ea)

Here is my original file: https://drive.google.com/file/d/1FswDATSIgQifnoTLTHWJT6FZfSptczV4/view?usp=sharing

**Comment by FelipeDefensor** [comment]
I managed to restore the file to valid state. Please check if you can open this: [Bruch Skizze_recovered.zip](https://github.com/user-attachments/files/17384206/Bruch.Skizze_recovered.zip)

**Comment by m-k94** [comment]
That file works fine, thanks!

**Comment by johentsch** [comment]
Ran into the same problem with the same piece. Cannot recall the exact sequence of events but I did store these two error messages:

```python-traceback
File "tilia\ui\actions.py", line 139, in callback
File "tilia\requests\post.py", line 239, in post
File "tilia\undo_manager.py", line 72, in undo
File "tilia\requests\post.py", line 239, in post
File "tilia\app.py", line 110, in on_restore_state
File "tilia\timelines\collection\collection.py", line 287, in restore_state
File "tilia\timelines\collection\collection.py", line 300, in _restore_timeline_state
File "tilia\timelines\base\timeline.py", line 191, in clear
File "tilia\timelines\base\timeline.py", line 396, in clear
File "tilia\timelines\hierarchy\timeline.py", line 602, in delete_component
File "tilia\timelines\hierarchy\timeline.py", line 622, in _update_genealogy_after_deletion
AttributeError: 'int' object has no attribute 'parent'
```

and:

```python-traceback
File "tilia\main.py", line 9, in <module>
File "tilia\main.py", line 5, in main
File "tilia\boot.py", line 52, in boot
File "tilia\file\file_manager.py", line 192, in open
File "tilia\requests\post.py", line 239, in post
File "tilia\app.py", line 155, in on_file_load
File "tilia\timelines\collection\collection.py", line 279, in deserialize_timelines
File "tilia\timelines\collection\collection.py", line 197, in create_timeline
File "tilia\timelines\base\timeline.py", line 203, in deserialize_components
File "tilia\timelines\hierarchy\timeline.py", line 137, in deserialize_components
File "tilia\timelines\base\timeline.py", line 402, in deserialize_components
File "tilia\timelines\serialize.py", line 89, in deserialize_components
File "tilia\timelines\serialize.py", line 153, in _substitute_ids_for_reference_to_components
File "tilia\timelines\serialize.py", line 153, in <listcomp>
KeyError: 134
```

**Comment by FelipeDefensor** [comment]
Sorry to hear that. Were you also trying to undo something? If that's the case, the intended fix for Maik's problem should also prevent whatever happened here.

Thanks for the error message. There's an `int` where a reference to a `Hierarchy` should be. I will look through the code to see where that may be happening.

In case that broke the file, I can probably restore if you send it to me.

**Comment by FelipeDefensor** [comment]
@azfoo just so you are aware, I think those kinds of errors should make us rethink how to handle the `children` and `parent` attributes. There have been a lot of bugs related to wrong types in those attributes. I think they should all be `int`s, as those are easily traced to the target object and have the significant advatage of being serializable, avoiding a lot of complexity when storing state or saving/opening files.

**Comment by FelipeDefensor** [comment]
I'm pretty sure this was caused by hierarchy duplication at some point, although it is hard to say for sure. Commit ab775e0 checks for hierarchy duplication when instatiating them and #223 prevents going to an invalid state by undoing, so I'm betting this has been fixed.

**Comment by FelipeDefensor** [comment]
The linked pull request that fixes this (#223) has been merged manually.
